## Loading the standard libraries

In [1]:
import pandas as pd
import numpy as np

## Loading the data

In [2]:
data = pd.read_csv('Sachin_ODI.csv')
data.head()

runs  NotOut mins  bf  fours  sixes      sr  Inns           Opp    Ground  \
0    13       0   30  15      3      0   86.66     1   New Zealand    Napier   
1    37       0   75  51      3      1   72.54     2  South Africa  Hamilton   
2    47       0   65  40      7      0  117.50     2     Australia   Dunedin   
3    48       0   37  30      9      1  160.00     2    Bangladesh   Sharjah   
4     4       0   13   9      1      0   44.44     2      Pakistan   Sharjah   

         Date        Winner    Won  century  
0  1995-02-16   New Zealand  False    False  
1  1995-02-18  South Africa  False    False  
2  1995-02-22         India   True    False  
3  1995-04-05         India   True    False  
4  1995-04-07      Pakistan  False    False

In [3]:
data.shape

(360, 14)

## Observations:

- We are looking at 360 ODI matches played by Sachin
- sr means strike rate
- bf means balls faced

In [4]:
data = data.drop(['Ground', 'Date'], axis = 1)
data.head()

runs  NotOut mins  bf  fours  sixes      sr  Inns           Opp  \
0    13       0   30  15      3      0   86.66     1   New Zealand   
1    37       0   75  51      3      1   72.54     2  South Africa   
2    47       0   65  40      7      0  117.50     2     Australia   
3    48       0   37  30      9      1  160.00     2    Bangladesh   
4     4       0   13   9      1      0   44.44     2      Pakistan   

         Winner    Won  century  
0   New Zealand  False    False  
1  South Africa  False    False  
2         India   True    False  
3         India   True    False  
4      Pakistan  False    False

1. If total matches played in the dataset is 360, it means that the total Sample Space = 360

## How to get the Sample Space of the data:

In [5]:
data.shape[0]

360

In [6]:
len(data)

360

## Event A : India winning the match

In [7]:
## Getting the sample space for event A

data['Won'].value_counts()

True     184
False    176
Name: Won, dtype: int64

1. The Sample Space for India winning the match is 184

## Probability of Event A :

In [8]:
## Basic Approach or manual approach

184 / 360

0.5111111111111111

In [9]:
## Pandas Approach

df_won = data[data['Won'] == True]
df_won.head()

runs  NotOut mins   bf  fours  sixes      sr  Inns          Opp Winner  \
2    47       0   65   40      7      0  117.50     2    Australia  India   
3    48       0   37   30      9      1  160.00     2   Bangladesh  India   
5   112       1  137  107     15      1  104.67     2    Sri Lanka  India   
6    41       0   51   41      5      0  100.00     2    Sri Lanka  India   
8    39       0   79   51      4      0   76.47     2  New Zealand  India   

    Won  century  
2  True    False  
3  True    False  
5  True     True  
6  True    False  
8  True    False

In [10]:
df_won.shape

(184, 12)

In [11]:
p_a = len(df_won) / len(data)
p_a

0.5111111111111111

## Observation:

- 0.51 (probability) represents that 50% is the winning rate for India

## Probability of losing

In [12]:
1 - p_a

0.48888888888888893

## Observation:

- Probability of losing is negative of Probability of winning
- P(A') = 1 - P(A)                [A' means A complement]
- 0.49 (probability) indicates that 49% is the losing rate for India

In [15]:
### Probability of losing using Pandas Approach

df_lose = data[data['Won'] == False]
df_lose

runs  NotOut mins   bf  fours  sixes      sr  Inns           Opp  \
0      13       0   30   15      3      0   86.66     1   New Zealand   
1      37       0   75   51      3      1   72.54     2  South Africa   
4       4       0   13    9      1      0   44.44     2      Pakistan   
7      30       0   30   20      5      0  150.00     1   New Zealand   
10     65       0   96   59      9      1  110.16     2   New Zealand   
..    ...     ...  ...  ...    ...    ...     ...   ...           ...   
352    15       0   25   24      2      0   62.50     2     Sri Lanka   
353     3       0   21   12      0      0   25.00     2     Australia   
354    22       0   33   23      3      0   95.65     2     Sri Lanka   
355    14       0   34   15      2      0   93.33     2     Australia   
358   114       0  205  147     12      1   77.55     1    Bangladesh   

           Winner    Won  century  
0     New Zealand  False    False  
1    South Africa  False    False  
4        Pakistan  False    False  
7     New Zealand  False    False  
10    New Zealand  False    False  
..            ...    ...      ...  
352          tied  False    False  
353     Australia  False    False  
354     Sri Lanka  False    False  
355     Australia  False    False  
358    Bangladesh  False     True  

[176 rows x 12 columns]

In [16]:
len(df_lose)

176

In [18]:
p_lose = len(df_lose) / len(data)
p_lose

0.4888888888888889

## Event B : Sachin Scoring a Century

In [19]:
### Sample Space for Sachin Scoring Century

data['century'].value_counts()

False    314
True      46
Name: century, dtype: int64

In [20]:
p_cen = 46 / 360
p_cen

0.12777777777777777

## Observations:

- 12.77 ~ 13% is the century rate for Sachin.

## Pandas Approach for Event B

In [21]:
df_century = data[data['century'] == True]
df_century

runs  NotOut mins   bf  fours  sixes      sr  Inns           Opp  \
5     112       1  137  107     15      1  104.67     2     Sri Lanka   
12    127       1  175  138     15      1   92.02     2         Kenya   
15    137       0  198  137      8      5  100.00     1     Sri Lanka   
20    100       0    -  111      9      1   90.09     1      Pakistan   
23    118       0    -  140      8      2   84.28     1      Pakistan   
27    110       0    -  138      5      1   79.71     1     Sri Lanka   
41    114       0  157  126     14      0   90.47     1  South Africa   
47    104       0  121   97      8      1  107.21     2      Zimbabwe   
55    117       0  188  137     13      2   85.40     2   New Zealand   
85    100       0  132   89      5      7  112.35     2     Australia   
91    143       0  187  131      9      5  109.16     2     Australia   
92    134       0  195  131     12      3  102.29     2     Australia   
95    100       1  156  103     13      0   97.08     2         Kenya   
99    128       0  182  131      8      2   97.70     1     Sri Lanka   
101   127       1  155  130     13      1   97.69     2      Zimbabwe   
104   141       0  191  128     13      3  110.15     1     Australia   
107   118       1  173  112     14      2  105.35     2      Zimbabwe   
110   124       1  140   92     12      6  134.78     2      Zimbabwe   
116   140       1  114  101     16      3  138.61     1         Kenya   
123   120       0  191  141     11      2   85.10     1     Sri Lanka   
128   186       1  221  150     20      3  124.00     1   New Zealand   
143   122       0  195  138     12      0   88.40     2  South Africa   
155   101       0    -  140      3      1   72.14     1     Sri Lanka   
162   146       0  200  153     15      2   95.42     1      Zimbabwe   
167   139       0  197  125     19      0  111.20     1     Australia   
173   122       1  210  131     12      1   93.12     2   West Indies   
175   101       0  209  129      9      0   78.29     1  South Africa   
179   146       0  186  132     17      0  110.60     1         Kenya   
191   105       1  167  108      8      1   97.22     1       England   
194   113       0  141  102     12      1  110.78     1     Sri Lanka   
206   152       0  193  151     18      0  100.66     1       Namibia   
215   100       0  158  119      9      1   84.03     1     Australia   
219   102       0  124   91     12      1  112.08     1   New Zealand   
229   141       0  173  135     17      1  104.44     2      Pakistan   
245   123       0  192  130     12      2   94.61     1      Pakistan   
258   100       0  188  113     10      1   88.49     1      Pakistan   
262   141       1    -  148     13      5   95.27     1   West Indies   
276   100       1  110   76     10      1  131.57     1   West Indies   
313   117       1  235  120     10      0   97.50     2     Australia   
322   163       1  206  133     16      5  122.55     1   New Zealand   
325   138       0  214  133     10      1  103.75     1     Sri Lanka   
331   175       0  210  141     19      4  124.11     2     Australia   
338   200       1  226  147     25      3  136.05     1  South Africa   
342   120       0  170  115     10      5  104.34     1       England   
345   111       0  184  101      8      3  109.90     1  South Africa   
358   114       0  205  147     12      1   77.55     1    Bangladesh   

           Winner    Won  century  
5           India   True     True  
12          India   True     True  
15      Sri Lanka  False     True  
20       Pakistan  False     True  
23          India   True     True  
27      Sri Lanka  False     True  
41          India   True     True  
47          India   True     True  
55          India   True     True  
85          India   True     True  
91      Australia  False     True  
92          India   True     True  
95          India   True     True  
99          India   True     True  
101         India   True     True

In [22]:
len(df_century)

46

In [24]:
p_century = len(df_century) / len(data)
p_century

0.12777777777777777

## Probability for sachin not scoring a century

In [25]:
1 - p_century

0.8722222222222222

## Probability of India Winning and Sachin Scoring a century

In [26]:
data.head()

runs  NotOut mins  bf  fours  sixes      sr  Inns           Opp  \
0    13       0   30  15      3      0   86.66     1   New Zealand   
1    37       0   75  51      3      1   72.54     2  South Africa   
2    47       0   65  40      7      0  117.50     2     Australia   
3    48       0   37  30      9      1  160.00     2    Bangladesh   
4     4       0   13   9      1      0   44.44     2      Pakistan   

         Winner    Won  century  
0   New Zealand  False    False  
1  South Africa  False    False  
2         India   True    False  
3         India   True    False  
4      Pakistan  False    False

In [28]:
df_iwsc = data[(data['Won'] == True) & (data['century'] == True)]
df_iwsc

runs  NotOut mins   bf  fours  sixes      sr  Inns           Opp Winner  \
5     112       1  137  107     15      1  104.67     2     Sri Lanka  India   
12    127       1  175  138     15      1   92.02     2         Kenya  India   
23    118       0    -  140      8      2   84.28     1      Pakistan  India   
41    114       0  157  126     14      0   90.47     1  South Africa  India   
47    104       0  121   97      8      1  107.21     2      Zimbabwe  India   
55    117       0  188  137     13      2   85.40     2   New Zealand  India   
85    100       0  132   89      5      7  112.35     2     Australia  India   
92    134       0  195  131     12      3  102.29     2     Australia  India   
95    100       1  156  103     13      0   97.08     2         Kenya  India   
99    128       0  182  131      8      2   97.70     1     Sri Lanka  India   
101   127       1  155  130     13      1   97.69     2      Zimbabwe  India   
104   141       0  191  128     13      3  110.15     1     Australia  India   
107   118       1  173  112     14      2  105.35     2      Zimbabwe  India   
110   124       1  140   92     12      6  134.78     2      Zimbabwe  India   
116   140       1  114  101     16      3  138.61     1         Kenya  India   
123   120       0  191  141     11      2   85.10     1     Sri Lanka  India   
128   186       1  221  150     20      3  124.00     1   New Zealand  India   
143   122       0  195  138     12      0   88.40     2  South Africa  India   
167   139       0  197  125     19      0  111.20     1     Australia  India   
173   122       1  210  131     12      1   93.12     2   West Indies  India   
179   146       0  186  132     17      0  110.60     1         Kenya  India   
194   113       0  141  102     12      1  110.78     1     Sri Lanka  India   
206   152       0  193  151     18      0  100.66     1       Namibia  India   
215   100       0  158  119      9      1   84.03     1     Australia  India   
219   102       0  124   91     12      1  112.08     1   New Zealand  India   
276   100       1  110   76     10      1  131.57     1   West Indies  India   
313   117       1  235  120     10      0   97.50     2     Australia  India   
322   163       1  206  133     16      5  122.55     1   New Zealand  India   
325   138       0  214  133     10      1  103.75     1     Sri Lanka  India   
338   200       1  226  147     25      3  136.05     1  South Africa  India   

      Won  century  
5    True     True  
12   True     True  
23   True     True  
41   True     True  
47   True     True  
55   True     True  
85   True     True  
92   True     True  
95   True     True  
99   True     True  
101  True     True  
104  True     True  
107  True     True  
110  True     True  
116  True     True  
123  True     True  
128  True     True  
143  True     True  
167  True     True  
173  True     True  
179  True     True  
194  True     True  
206  True     True  
215  True     True  
219  True     True  
276  True     True  
313  True     True  
322  True     True  
325  True     True  
338  True     True

In [30]:
## Sample space for India winning and sachin scoring the century

len(df_iwsc)

30

In [32]:
## Prob

p_iwsc = len(df_iwsc) / len(data)
p_iwsc

0.08333333333333333

## Observations:

- 8.3% of times India wins when Sachin scores century

## Probability of India Winning and Sachin Scoring a century using a simple approach

In [34]:
pd.crosstab(index = data['century'], columns = data['Won'], margins = True)

Won      False  True  All
century                  
False      160   154  314
True        16    30   46
All        176   184  360

In [35]:
30 / 360

0.08333333333333333

## Find the probability of India Winning given that Sachin has already scored a century in that match

A : Sachin Scoring Century   
B : India winning the match

P(B | A) ==> P(India winning the match | Sachin has already scored century)

In [36]:
pd.crosstab(index = data['century'], columns = data['Won'], margins = True)

Won      False  True  All
century                  
False      160   154  314
True        16    30   46
All        176   184  360

Sample Space (Sachin scoring a century) = 46  
Sample Space (India Winning out of 46 matches) = 30


Prob(B | A) = 30 / 46

In [37]:
p_b_a = 30 /46
p_b_a

0.6521739130434783

## Observations:

- 65% of times india wins a match when sachin has scored a century

## Probability of Sachin Scoring a century given that India has already lost that match

Event A : India Losing the match  
Event B : Sachin scoring a century

P(B | A) ==> P(Sachin Scoring a century | India has lost the match)